### ISSUE TITLE
Introduce TestContext.addWatchFile or TestContext.readFile

### ISSUE THREAD

**Issue by gtm-nayan** [vitest-dev/vitest#3378]
### Clear and concise description of the problem

I have some tests that read files. When those files change I want to rerun those tests. There is a config option that allows registering files to be watched but that's all or nothing, it redoes all the tests when any of the files change.

### Suggested solution

Introduce functions to the test context object,
1. addWatchFile: register a file to be watched for the current test
2. readFile: wrapper around fs.readFile that resolves paths relative to the current test file and also registers the file to be watched for the current test

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by phawxby** [comment]
@gtm-nayan I've got a similar issue but I can't see any config option that allows for registering additional files to be watch. Am I being blind?

**Comment by sheremet-va** [comment]
> @gtm-nayan I've got a similar issue but I can't see any config option that allows for registering additional files to be watch. Am I being blind?

Files to watch are added automatically when imported. If you need to rerun every test, use [`forceRerunTriggers`](https://vitest.dev/config/#forcereruntriggers) option.

**Comment by phawxby** [comment]
@sheremet-va that's what I thought but I'm not sure if it's me misunderstanding the docs or it's not working. For background I have a bunch of json files I use to build out the test suite, I would like a change to those json files to trigger a full re-run.

```
{
  test: {
    include: [ '**/*.test.{js,mjs,cjs,ts,mts,cts,jsx,tsx}' ],
    forceRerunTriggers: [
      '**/package.json/**',
      '**/{vitest,vite}.config.*/**',
      '**/*.json'
    ],
    watchExclude: [ '**/node_modules/**', '**/dist/**' ]
  }
```

Above is a console.log of the export from `vitest.config.ts`. I realise `**/*.json` is a bit greedy but I'm debugging. File paths are like `./src/foo.json`

**Comment by marlass** [comment]
I've similar problem. I've a bunch of files that are used as examples to run tests against and it's not efficient to manually run all tests in watch mode when changing those files. 

I was expecting `forceRerunTriggers` to work in watch mode. Watching these files like regular test files and rerunning all of the tests when those changes (similarly like when vitest config is changed)

**Comment by sheremet-va** [comment]
> I was expecting `forceRerunTriggers` to work in watch mode. Watching these files like regular test files and rerunning all of the tests when those changes (similarly like when vitest config is changed)

There is a bug in the latest version that broke `forceRerunTriggers`. It will be fixed in 0.34.0.
